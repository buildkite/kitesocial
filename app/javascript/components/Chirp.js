import * as React from 'react';
import PropTypes from 'prop-types';

import UserLink from './UserLink';
import RelativeDateTime from './RelativeDateTime';

export default class Chirp extends React.PureComponent {
  static propTypes = {
    chirp: PropTypes.shape({
      content: PropTypes.string.isRequired,
      author: PropTypes.object.isRequired,
      mentions: PropTypes.arrayOf(
        PropTypes.object
      ).isRequired,
      created_at: PropTypes.string.isRequired,
      liked: PropTypes.bool.isRequired,
      like_url: PropTypes.string.isRequired
    }).isRequired
  };

  state = {
    liking: false,
    likeOverride: null,
    likesCountOverride: null
  };

  get liked() {
    if (this.state.likeOverride !== null) {
      return this.state.likeOverride;
    }

    return this.props.chirp.liked;
  }

  renderLikeButton() {
    const { chirp: { like_url } } = this.props;

    return (
      <a
        href={like_url}
        onClick={this.handleLikeButtonClick}
        disabled={this.state.liking}
      >
        {this.liked ? 'Unlike' : 'Like'}
      </a>
    );
  }

  handleLikeButtonClick = (event) => {
    event.preventDefault();

    if (this.state.liking) {
      return;
    }

    const { chirp: { like_url } } = this.props;

    this.setState({ liking: true }, () => {
      fetch(
        `${like_url}.json`,
        {
          headers: {
            'X-CSRF-Token': document.querySelector('meta[name="csrf-token"]').content
          },
          method: this.liked ? 'DELETE' : 'POST'
        }
      )
        .then((response) => {
          if (response.ok) {
            return response.json();
          }
          this.setState({ liking: false });

        })
        .then((chirp) => {
          this.setState({
            liking: false,
            likeOverride: chirp.liked,
            likesCountOverride: chirp.likes_count
          });
        });
    });
  };
  
  renderLikesCount() {
    const likes = this.state.likesCountOverride !== null ? this.state.likesCountOverride : this.props.chirp.likes_count;

    return `${likes} like${likes === 1 ? '' : 's'}`;
  }

  render() {
    const { chirp } = this.props;

    const knownMentions = chirp.mentions.reduce((accumulator, value) => {
      accumulator[value.name.toLowerCase()] = value;
      return accumulator;
    }, {});

    // JS can't do negative lookbehinds yet (https://caniuse.com/mdn-javascript_builtins_regexp_lookbehind_assertion),
    // so we need more manual checks in how we process mentions into links
    const output = chirp.content.split(/(@\w+)/).map((segment, index, array) => {
      if ((index === 0 || array[index - 1].match(/[^\w]$/)) && segment.startsWith('@')) {
        const slicedSegment = segment.slice(1).toLowerCase();

        if (Object.prototype.hasOwnProperty.call(knownMentions, slicedSegment)) {
          return <UserLink user={knownMentions[slicedSegment]} key={index} />; // eslint-disable-line react/no-array-index-key
        }
      }

      return segment;
    });

    return (
      <div>
        <p>{ output }</p>
        -- <UserLink user={chirp.author} />
        {', '}
        <RelativeDateTime value={chirp.created_at} />
        {' • '}
        {this.renderLikesCount()}
        {' • '}
        {this.renderLikeButton()}
      </div>
    );
  }
}
