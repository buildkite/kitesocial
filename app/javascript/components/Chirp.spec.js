/* eslint-env jest */

import React from 'react';
import renderer from 'react-test-renderer';
import Chirp from './Chirp';

// RelativeDateTime changes based on the current time,
// mock it here and we'll test that in its own specs.
jest.mock('./RelativeDateTime', () => 'RelativeDateTime');

describe('mention tracking', () => {
  it('links to mentioned users', () => {
    const component = renderer.create(
      <Chirp
        chirp={{
          id: 1,
          author: { id: 1,
            name: "alice",
            url: "/users/1" },
          content: "hey @bob @bob @carol @bob @bob",
          created_at: "2021-04-22T01:08:10Z",
          mentions: [
            { id: 2,
              name: "bob",
              url: "/users/2" },
            { id: 3,
              name: "carol",
              url: "/users/3" }
          ],
          updated_at: "2021-04-22T01:08:10Z",
          likes_count: 0,
          liked: false,
          like_url: "/chirps/1/like"
        }}
      />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('avoids things that look like email addresses', () => {
    const component = renderer.create(
      <Chirp
        chirp={{
          id: 1,
          author: { id: 1,
            name: "alice",
            url: "/users/1" },
          content: "I just got a great email from tony@bob.com!",
          created_at: "2021-04-22T01:08:10Z",
          mentions: [],
          updated_at: "2021-04-22T01:08:10Z",
          likes_count: 0,
          liked: false,
          like_url: "/chirps/1/like"
        }}
      />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
