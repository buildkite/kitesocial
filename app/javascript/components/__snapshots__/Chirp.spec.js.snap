// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mention tracking avoids things that look like email addresses 1`] = `
<div>
  <p>
    I just got a great email from tony
    @bob
    .com!
  </p>
  -- 
  <a
    href="/users/1"
  >
    @
    alice
  </a>
  , 
  <RelativeDateTime
    value="2021-04-22T01:08:10Z"
  />
</div>
`;

exports[`mention tracking links to mentioned users 1`] = `
<div>
  <p>
    hey 
    <a
      href="/users/2"
    >
      @
      bob
    </a>
     
    <a
      href="/users/2"
    >
      @
      bob
    </a>
     
    <a
      href="/users/3"
    >
      @
      carol
    </a>
     
    <a
      href="/users/2"
    >
      @
      bob
    </a>
     
    <a
      href="/users/2"
    >
      @
      bob
    </a>
    
  </p>
  -- 
  <a
    href="/users/1"
  >
    @
    alice
  </a>
  , 
  <RelativeDateTime
    value="2021-04-22T01:08:10Z"
  />
</div>
`;
